### PR TITLE
Add support for atomic wait and wake operators

### PR DESCRIPTION
**Pull request by dschuff**

### CODE REVIEW

**Comment by dschuff** [comment]
Actually wait, this PR has the wrong type for atomic.wait. The naming is also misleanding. I'm going to file an issue against the spec proposal before landing this.

**Comment by dschuff** [comment]
OK, the types are now fixed. We can quibble with the spec later if we want but this PR should work as-is.

File: src/ast/effects.h [hunk 1]
```diff
@@ -204,6 +204,24 @@ struct EffectAnalyzer : public PostWalker<EffectAnalyzer> {
     isAtomic = true;
     if (!ignoreImplicitTraps) implicitTrap = true;
   }
+  void visitAtomicWait(AtomicWait* curr) {
+    readsMemory = true;
+    // AtomicWait doesn't strictly write memory, but it does modify the waiters
```
Comment (kripken) [inline src/ast/effects.h]: this comment should go one line up to match the function below

File: src/ast/effects.h [hunk 2]
```diff
@@ -204,6 +204,24 @@ struct EffectAnalyzer : public PostWalker<EffectAnalyzer> {
...
+    // list associated with the specified address, which we can think of as a
+    // write.
+    writesMemory = true;
+    isAtomic = true;
+    if (!ignoreImplicitTraps) implicitTrap = true;
```
Comment (kripken) [inline src/ast/effects.h]: i haven't read the spec, how can these trap? seems odd that wait/wake could trap...

File: src/ast/effects.h (hunk 1 above)
Comment (kripken) [inline src/ast/effects.h]: or vice versa

File: src/ast/effects.h (hunk 2 above)
Comment (dschuff) [inline src/ast/effects.h]: Wait and wake take addresses in linear memory (like using the memory address for the value of a mutex), so they can trap for the same reasons as other memory operations. See https://github.com/WebAssembly/threads/blob/master/proposals/threads/Overview.md#wait-and-wake-operators

**Pull request merged**